### ISSUE TITLE
How to use index funds in optimizer?

### ISSUE THREAD

**Issue by JohnMansell** [robertmartin8/PyPortfolioOpt#125]
**What are you trying to do?*
I am writing a portfolio optimizer for work. As a base line, I am optimizing the portfolios with the built in options for PyPortfolioOpt. I will then build some of our own risk models into PyPfOpt and compare results.

**What data are you using?**
I have so far run about 100 portfolios. Each portfolio has between 5 - 15 assets. The data is date, asset, closing price for each asset. I've done my best to format the data to match the example data.
I've saved two example portfolios to CSV and included them in a zip file below.

**Problem**
Roughly 50% of my portfolios fail at the opt.solve step. They fail with the error : 
> Problem does not follow DCP rules. Specifically: 
> The objective is not DCP. Its following subexpressions are not:
> QuadForm(var116, [[ 0.04810146  0.04616764  0.03885335 -0.00271883 -0.00106712 -0.00485717]
 > [ 0.04616764  0.04726095  0.02809275 -0.00269661 -0.00101515 -0.00486945]
 > [ 0.03885335  0.02809275  0.02448724 -0.00228728 -0.00094065 -0.00394085]
 > [-0.00271883 -0.00269661 -0.00228728  0.00302117  0.00119272  0.00473196]
 > [-0.00106712 -0.00101515 -0.00094065  0.00119272  0.00063444  0.00171789]
 > [-0.00485717 -0.00486945 -0.00394085  0.00473196  0.00171789  0.00913102]])

The portfolios which fail all seem to include index funds (IRX, SPX, IVV etc). Portfolios which don't include any index are solvable.

I have tried different solvers, (SCS, ECOS, OSQP) as well as using max_sharpe and min_volatility. Neither of these changed the success or failure of the optimizer.

**Questions**
- Is there a known reason why index funds would break the optimizer?
- Is there something about the index funds which make the problem no longer convex?
- Is there a way I should be sanitizing my data?
- This is probably a user error. I know that the problems are most often from users not formatting data correctly. Since its working on half of the portfolios, I'm not sure where the error is coming from, so I don't know how to change the data formatting.

``` python
import pandas as pd

from pypfopt import risk_models
from pypfopt import expected_returns
from pypfopt import EfficientFrontier


# -----------------------------
#       Optimizer
# -----------------------------
def optimizer(df):
    # --- Returns and Risk
    returns = df.pct_change().dropna()
    mu = expected_returns.mean_historical_return(df)
    S = risk_models.sample_cov(df)

    # --- Max Sharpe
    ef = EfficientFrontier(mu, S)
    weights = ef.max_sharpe()
    ef.portfolio_performance(verbose=True)


# --- Import Data
df_748 = pd.read_csv("portfolio_748.csv", parse_dates=True, index_col="date")
df_240 = pd.read_csv("portfolio_240.csv", parse_dates=True, index_col="date")

# --- Sanity Check
print(df_240)
print(df_748)

# --- Run Optimizer
optimizer(df_748)  # Success
optimizer(df_240)  # Fails
```

[example_data.zip](https://github.com/robertmartin8/PyPortfolioOpt/files/4656849/example_data.zip)

**Comment by robertmartin8** [comment]
Hi John,

Thanks for raising the issue, and for providing some sample data! 

Unfortunately, I wasn't able to reproduce the error when I ran it locally – it managed to produce a portfolio without any problems:

```python
              MDY   NAESX      SPX  VBIIX  VBISX  VBLTX
date                                                    
1950-01-03     NaN     NaN    16.66    NaN    NaN    NaN
1950-01-04     NaN     NaN    16.85    NaN    NaN    NaN
1950-01-05     NaN     NaN    16.93    NaN    NaN    NaN
1950-01-06     NaN     NaN    16.98    NaN    NaN    NaN
1950-01-09     NaN     NaN    17.08    NaN    NaN    NaN
...            ...     ...      ...    ...    ...    ...
2020-05-13  371.56  161.82  2819.98  41.75  26.83  56.68
2020-05-14  375.74  163.39  2852.48  41.78  26.83  56.68
2020-05-15  378.80  165.15  2863.68  41.78  26.83  56.68
2020-05-18  401.75  174.34  2953.89  41.69  26.80  56.68
2020-05-19  396.22  171.95  2922.92  41.85  26.83  56.68

[17709 rows x 6 columns]
                 F       GE     GM    IDV    TSLA
date                                             
1962-01-02     NaN    74.75    NaN    NaN     NaN
1962-01-03     NaN    73.99    NaN    NaN     NaN
1962-01-04     NaN    73.12    NaN    NaN     NaN
1962-01-05     NaN    71.25    NaN    NaN     NaN
1962-01-08     NaN    71.12    NaN    NaN     NaN
...            ...      ...    ...    ...     ...
2020-05-13  192.37   963.27  23.13  35.55  790.92
2020-05-14  199.30   948.30  24.05  35.27  803.29
2020-05-15  199.71   913.36  24.39  35.20  799.13
2020-05-18  216.42  1043.13  26.74  36.74  813.59
2020-05-19  216.01  1033.15  26.61  35.88  807.97

[14696 rows x 5 columns]
Expected annual return: 50.6%
Annual volatility: 54.8%
Sharpe Ratio: 0.89
Expected annual return: 4.7%
Annual volatility: 2.9%
Sharpe Ratio: 0.93
```

One way for verifying whether it's PyPortfolioOpt or some environment issue is to try out the same function in Binder – on the main Readme, if you click "Launch Binder", it will create a cloud environment using the correct dependencies. Then, you can upload your two csv files, create a Jupyter notebook, and run the test code. When I did this, it worked, although it warned that some of the eigenvalues are negative.

So what I suspect is happening is that you are on a slightly older version of pypfopt (I only recently introduced an automatic function to detect non-PSD covariance matrices and repair them), so you might have some success with `pip install -U pyportfolioopt`

If that doesn't work, could you let me know:

- Your operating system
- Python version
- PyPortfolioOpt version
- cvxpy, numpy, pandas versions.

Best,
Robert

**Comment by JohnMansell** [comment]
Thank you for your quick response. The project is now working.

I tried it on Binder and confirmed that it was working. I then created a new python environment, copied the project over, and installed all the dependencies. The submitted code is now working appropriately. 

I'm not sure which dependency was broken, or if I accidentally broke something in the PyPortfolioOptimizer when I was testing and prodding the various parts of it.

Thank you very much for your help.
